File: Julia/Other/anagrams.jl
function sort_word(str)
  return join(sort(collect(str)))
end

function get_anagrams(file, hash_function)
  anagrams = Dict()
  open(file,"r") do f
      for line in eachline(f)
          word = chomp(strip(line))
          hashed_key = hash_function(word)

          if haskey(anagrams, hashed_key)
            push!(anagrams[hashed_key],word)
          else
            anagrams[hashed_key] = [word]
          end
      end
  end
  return anagrams
end

fname = "/Users/ioanwilliams/github/furry-palm-tree/words.txt"
anagrams = get_anagrams(fname, sort_word)

sorted = []
for v in values(anagrams)
    push!(sorted, (length(v),v))
end

#sort by first element of tuple
sort!(sorted, by = x -> x[1])

for (a,b) in sorted
  println(a,": ",join(b,","))
end

File: anagrams.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Jan 12 21:40:30 2017

@author: ioanwilliams
"""

def sort_word(s):
    t = list(s).sort()
    t = ''.join(t)
    return t


def all_anagrams(filename):
    anagrams = {}
    for line in open(filename):
        word = line.strip().lower()
        sorted_word = sort_word(word)

        if sorted_word not in anagrams:
            anagrams[sorted_word] = [word]
        else:
            anagrams[sorted_word].append(word)
    return anagrams


def print_anagram_sets(d):
    for v in d.values():
        if len(v) > 1:
            print(len(v), v)


def print_anagram_sets_in_order(d):
    t = []
    for v in d.values():
        if len(v) > 1:
            t.append((len(v), v))

    t.sort()

    for x in t:
        print(x)


if __name__ == '__main__':
    anagram_map = all_anagrams('/Users/ioanwilliams/Documents/words.txt')
    print_anagram_sets_in_order(anagram_map)